File: zul/src/org/zkoss/zul/api/Spinner.java
/* Spinner.java

{{IS_NOTE
	Purpose:
		
	Description:
		
	History:
		Tue Oct 22 14:45:31     2008, Created by Flyworld
}}IS_NOTE

Copyright (C) 2008 Potix Corporation. All Rights Reserved.

{{IS_RIGHT
	This program is distributed under GPL Version 3.0 in the hope that
	it will be useful, but WITHOUT ANY WARRANTY.
}}IS_RIGHT
 */
package org.zkoss.zul.api;

import org.zkoss.zk.ui.WrongValueException;

/**
 * An edit box for holding a constrained integer.
 * 
 * <p>
 * Default {@link #getZclass}: z-spinner.
 * 
 * <p>
 * spinner supports below key events. <lu>
 * <li>0-9 : set the value on the inner text box.
 * <li>delete : clear the value to empty (null) </lu>
 * 
 * @author gracelin
 * @since 3.5.2
 */
public interface Spinner extends org.zkoss.zul.impl.api.NumberInputElement {

	/**
	 * Returns the value (in Integer), might be null unless a constraint stops
	 * it.
	 * 
	 * @exception WrongValueException
	 *                if user entered a wrong value
	 */
	public Integer getValue() throws WrongValueException;

	/**
	 * Returns the value in int. If null, zero is returned.
	 */
	public int intValue() throws WrongValueException;

	/**
	 * Sets the value (in Integer).
	 * 
	 * @exception WrongValueException
	 *                if value is wrong
	 */
	public void setValue(Integer value) throws WrongValueException;

	/**
	 * Return the step of spinner
	 */
	public int getStep();

	/**
	 * Set the step of spinner
	 */
	public void setStep(int step);

	/**
	 * Returns whether the button (on the right of the textbox) is visible.
	 * <p>
	 * Default: true.
	 */
	public boolean isButtonVisible();

	/**
	 * Sets whether the button (on the right of the textbox) is visible.
	 */
	public void setButtonVisible(boolean visible);

}

File: zul/src/archive/web/js/zul/inp/mold/spinner.js
/* spinner.js

	Purpose:
		
	Description:
		
	History:
		Thu May 27 10:17:24     2009, Created by kindalu

Copyright (C) 2009 Potix Corporation. All Rights Reserved.

This program is distributed under LGPL Version 2.1 in the hope that
it will be useful, but WITHOUT ANY WARRANTY.
*/
function (out) {
	var uuid = this.uuid,
		isButtonVisible = this._buttonVisible;
	
	out.push('<span', this.domAttrs_({text:true}), '>',
			'<input id="', uuid,'-real"', 'class="', this.$s('input'));

	if (!isButtonVisible)
		out.push(' ', this.$s('rightedge'));
	
	out.push('" autocomplete="off"',
			this.textAttrs_(),'/>', '<span id="', uuid,'-btn"',
			'class="', this.$s('button'));
	
	if (!isButtonVisible)
		out.push(' ', this.$s('disabled'));
		
	var iconClass = this.$s('icon') + ' ';
	out.push('"><a id="', uuid, '-btn-up" class="', iconClass, this.$s('up'),
			'"><i class="', this.getBtnUpIconClass_(), '"></i></a><i class="', this.$s('separator'),
			'"></i><a id="', uuid, '-btn-down" class="', iconClass, this.$s('down'),
			'"><i class="', this.getBtnDownIconClass_(), '"></i></a>');

	out.push('</span></span>');
}
